### DISCUSSION TITLE
Unable to insert data through EntityRepository.create or EntityRepository.persist

### DISCUSSION THREAD

**Discussion by Shivam-Pathak1** [mikro-orm/mikro-orm#1571] [Q&A] [answered]
Hello there, :)

I am trying to test my Entity operations using the code in the file.
I am creating a userRepository object as follows:
![image](https://user-images.githubusercontent.com/56876559/111469376-ea0a9080-8726-11eb-87dd-eca2ab214e2e.png)

When I console.log find{} from the repository, it fetches the previously stored records:
![image](https://user-images.githubusercontent.com/56876559/111469469-06a6c880-8727-11eb-8463-0d860ff29db5.png)
I create a dummy object using faker and it works fine but as soon as I try to create it in DB or persist it, it does not seem to work:
![image](https://user-images.githubusercontent.com/56876559/111469683-453c8300-8727-11eb-8bf9-b117ade57b22.png)

**Comment by B4nan** [comment, marked as answer]
Enable debug mode to see what queries are made, that should help a lot to understand what is happening. 

https://mikro-orm.io/docs/configuration/#debugging--logging

From that you should see there is no insert query, because you did not make the EM aware of your user entity. Calling `flush` will only flush the managed entities = those loaded from database via EM. Your user entity is new one, not managed - you need to persist it first to make the EM know about it:

```ts
      const existingUser = new User(
        faker.name.firstName(),
        faker.name.lastName(),
        faker.internet.email(),
        userId,
      );
      userRepository.persist(existingUser); // <-- this will make the EM aware, so the flush call can save the entity
      await userRepository.flush();
```

There are some other things you had wrong, e.g. calling `em.create()` with entity instance is wrong, the method takes a DTO and creates the instance for you - it is an alternative way to use constructors. Using it and ignoring the return value is a misusage for sure. Don't use both at the same time - it might have some sideeffects (it might be actually your problem, hard to say).  Just drop the `userRepository.create(existingUser);` like as a whole. 

Also this is not very good way to create issues, next time just copy the code inside a code block - no images, no attachements - those just make it harder to debug for others.

**Reply by Shivam-Pathak1** [reply]
Thank you for the response. I will use code blocks from now on.
Well, I already used 
`userRepository.persist(existingUser); `
But have been receiving the following error:

`
ValidationError: Trying to persist not discovered entity of type User. Entity with this name was discovered, but not the prototype you are passing to the ORM. If using EntitySchema, be sure to point to the implementation via `class`.

`

Working my way for the  debugger now.

**Reply by B4nan** [reply]
Yes, that means you are trying to work with an entity that was not discovered. 

If you enable debug mode as I suggested, you will see what entities were discovered. Common issue with this is that you run via `ts-node` but discover JS entiteis or vice versa - for `ts-node` you need to discover the TS files, and for `node` you need to discover JS files. 

How does your ORM config look like? Do you use folder based discovery?

Maybe this is also related: #1501

**Reply by B4nan** [reply]
Actually the error message is already pointing to the issue I am talking about: `Entity with this name was discovered, but not the prototype you are passing to the ORM.`

So it is definitely issue with trying to run via ts-node but discovering JS files (or vice versa). Adding `tsNode: true` to the config (but only for tests, not for production!!) might help (but depends on how your config looks like).

**Reply by Shivam-Pathak1** [reply]
```
import * as dotenv from 'dotenv';
import { Logger } from '@nestjs/common';
import { MikroOrmModuleSyncOptions } from '@mikro-orm/nestjs';
import { TsMorphMetadataProvider } from '@mikro-orm/reflection';

export const ROOT_DIR = __dirname + '/../../../../';
export const SRC_DIR = ROOT_DIR + 'src/';
let path: string;
switch (process.env.NODE_ENV) {
  case 'test':
    path = ROOT_DIR + `env/test.env`;
    break;
  case 'production':
    path = ROOT_DIR + `env/production.env`;
    break;
  default:
    path = ROOT_DIR + `env/development.env`;
    process.env.NODE_ENV = 'development';
}
dotenv.config({ path: path });
console.log(path); //  todo use Logger

const logger = new Logger('MikroORM');

logger.debug(
  'DATABASE_CONNECTION_STRING',
  process.env.DATABASE_CONNECTION_STRING,
);

const mikroOrmPostgresConfig = {
  migrations: {
    tableName: 'mikro_orm_migrations', // name of database table with log of executed transactions
    path: SRC_DIR + 'migrations/' + process.env.NODE_ENV, // path to the folder with migrations
    emit: 'ts', // migration generation mode
    pattern: /^[\w-]+\d+\.ts$/, // regex pattern for the migration files
    transactional: true, // wrap each migration in a transaction
    disableForeignKeys: true, // wrap statements with `set foreign_key_checks = 0` or equivalent
    allOrNothing: false, // wrap all migrations in master transaction
    dropTables: true, // allow to disable table dropping
    safe: false, // allow to disable table and column dropping
  },
  metadataProvider: TsMorphMetadataProvider,
  tsNode: process.env.NODE_DEV === 'true' ? true : false,
  clientUrl: process.env.DATABASE_CONNECTION_STRING,
  debug: true,
  logger: logger.log.bind(logger),

  entities: [ROOT_DIR + 'dist/entities/*.js'], // All other entities that are not part of a module need to be imported here, to avoid duplicate inclusion.
  entitiesTs: [ROOT_DIR + 'src/entities/*.ts'], // All other entities that are not part of a module need to be imported here, to avoid duplicate inclusion.
  type: 'postgresql' as const,
  cli: {
    migrationsDir: SRC_DIR + 'migrations/' + process.env.NODE_ENV,
  },
} as MikroOrmModuleSyncOptions;

export default mikroOrmPostgresConfig;

```
You were right. I am running it with NODE_ENV value as "test" and thus tsNode was false. Made it true and it worked.

**Reply by B4nan** [reply]
Good to hear that. 

Btw do you have any specific reason to disable `allOrNothing` in the migrations config? I can't imagine why you would want that. Having that disabled means if you migrate multiple steps together, and some of them fail, the whole migration will not fail. You can get into invalid state or even loose some data with having that feature disabled.

**Reply by Shivam-Pathak1** [reply]
Well, you are right about this as well :) It might have been something not working in the past and the need of the hour might have been just to make db migration work. We will change that.
